cmd/guacone/cmd/vulnerability.go

```diff
@@ -124,12 +125,96 @@ func printVulnInfo(ctx context.Context, gqlclient graphql.Client, t table.Writer
 	var path []string
 	var tableRows []table.Row
 
-	depVulnPath, depVulnTableRows, err := searchPkgViaHasSBOM(ctx, gqlclient, opts.searchString, opts.depth, opts.isPurl)
-	if err != nil {
-		logger.Fatalf("error searching via hasSBOM: %v", err)
+	if opts.isPurl {
+		depVulnPath, depVulnTableRows, err := searchForSBOMViaPkg(ctx, gqlclient, opts.searchString, opts.depth, true)
+		if err != nil {
+			logger.Fatalf("error searching via hasSBOM: %v", err)
+		}
+
+		path = append(path, depVulnPath...)
+		tableRows = append(tableRows, depVulnTableRows...)
+
+		if len(depVulnPath) == 0 {
+			pkgInput, err := helpers.PurlToPkg(opts.searchString)
+			if err != nil {
+				logger.Fatalf("failed to parse PURL: %v", err)
+			}
+
+			pkgQualifierFilter := []model.PackageQualifierSpec{}
+			for _, qualifier := range pkgInput.Qualifiers {
+				// to prevent https://github.com/golang/go/discussions/56010
+				qualifier := qualifier
+				pkgQualifierFilter = append(pkgQualifierFilter, model.PackageQualifierSpec{
+					Key:   qualifier.Key,
+					Value: &qualifier.Value,
+				})
+			}
+
+			pkgFilter := &model.PkgSpec{
+				Type:       &pkgInput.Type,
+				Namespace:  pkgInput.Namespace,
+				Name:       &pkgInput.Name,
+				Version:    pkgInput.Version,
+				Subpath:    pkgInput.Subpath,
+				Qualifiers: pkgQualifierFilter,
+			}
+
+			o, err := model.Occurrences(ctx, gqlclient, model.IsOccurrenceSpec{
+				Subject: &model.PackageOrSourceSpec{
+					Package: pkgFilter,
+				},
+			})
+			if err != nil {
+				logger.Fatalf("error querying for occurrences: %v", err)
+			}
+
+			depVulnPath, depVulnTableRows, err = searchForSBOMViaArtifact(ctx, gqlclient, o.IsOccurrence[0].Artifact.Id, opts.depth, false)
+			if err != nil {
+				logger.Fatalf("error searching for SBOMs via artifact: %v", err)
+			}
```

Comment:
the function is getting a bit long - i think we can refactor some of these inner conditionals into functions or at least with some helpers around going from search string to an occurrence.

Reply:
Yes, this function is getting kind of long, I have refactored the code into smaller functions. Thanks!